### DISCUSSION TITLE
Additional folder contents - new option 'appear only when search'

### DISCUSSION THREAD

**Discussion by RuSieg** [Hofknecht/SystemTrayMenu#283] [Q&A] [answered]
I would be happy about the possibility if only the own directories would be displayed and everything below could be hidden on request.
So I could create my own menu program by program without ballast.

Thanks and greetings
Klaus

**Comment by Hofknecht** [comment]
Hello @RuSieg 
Thank you for the feedback!
That sounds exciting, we need more details about what exactly you have in mind.
maybe you could describe it with a few screenshots. mark with red the content that should not be there. In this you can also show us how your windows folder looks like

What do you mean with "everything below could be hidden on request"

Maybe you can describe us a little more in screenshots. 

I'm just thinking, wouldn't it be possible to make shortcuts from the folders and just add the shortcuts into the main folder?

or a option similar 'Only files'?
![2022-01-04 22_34_17-Window](https://user-images.githubusercontent.com/52528841/148126872-4587c029-95dd-40f3-9c9a-8ca1cf200adf.png)

We definitely need something to hide the contents of the additional folders and that they only appear if you use the search.
But that is probably something completely different

**Comment by RuSieg** [comment]
Hi @Markus,
thanks for the quick reply and thank you for this program. I have tried several alternatives but SystemTrayMenu comes closest to my ideas.
Here's a screenshot of what I mean.
![image](https://user-images.githubusercontent.com/7941998/148184051-610b0007-0197-4fed-8093-861eb80a6d5b.png)

I have installed a lot of programs and am slowly losing track. The desktop is full and Windows 11 doesn't make it any easier for me.
Thank you and greetings
Klaus

**Comment by Hofknecht** [comment]
@RuSieg 
Thanks for the screenshot, that makes it more clear.

already now there would be the possibility to make shortcuts of these folders and put them into a root folder for the STM. If I understand you correctly, then you still want to find the files with the search. I am currently thinking that you can then add the files folder in the 'Folders' tab, then (e.g. also the 'only files' option and then) you would need the 'only when search' checkboxcolumn/function described on our internal roadmap.

Of course we could also make a general option and there e.g. a checkbox 'show all files only when search'. but that is very fixed and inflexible. e.g. I could imagine that someone would still want to see a script 'Shutdown' anyway

In principle, it can be implemented quickly. The only problem is that in the 'folders' tab page in this grid, there is hardly any space for another column.

I'll generate an issue from this, we'll implement the option so that with 'only files' and 'only search' you get files in the stm root folder that are only displayed when you search

**Comment by Hofknecht** [comment, marked as answer]
generated 
https://github.com/Hofknecht/SystemTrayMenu/issues/284

**Comment by RuSieg** [comment]
@Markus,
by default I only want to see the links I have placed behind the directory names e.B. PRG. But when I search for a program in the search window, it should also be displayed to me.
I should then be able to call it up directly or, that would be very good, put a link from it on the desktop, which I can then move to PRG, for example.

**Comment by markus** [comment]
@RuSieg you would have to mention @Hofknecht in order to ping him and not me.

**Reply by RuSieg** [reply]
ok

**Comment by RuSieg** [comment]
@Hofknecht 
by default I only want to see the links I have placed behind the directory names e.B. PRG. But when I search for a program in the search window, it should also be displayed to me.
I should then be able to call it up directly or, that would be very good, put a link from it on the desktop, which I can then move to PRG, for example.

**Comment by Hofknecht** [comment]
@RuSieg 
Danke für deine Antwort!
Wir sind gerade nicht ganz sicher ob wir alles verstehen und wir sehen du kommst aus Deutschland, daher formulieren wir es auf deutsch.
Wir sehen auf dem Screenshot, dein root Ordner ist 'TrayMenu'. Erstmal wird ja aktuell alles angezeigt, Dateien und Ordner, welche sich im Ordner 'TrayMenu' befinden. 
Der Ordner 'TrayMenu'  kann auf den Desktop, aber ich persönlich habe meinen root Ordner unter D:\Seafile\Tools, da Seafile meine Cloud ist, damit alles über das Netzwerk gesichert wird. Der Desktop bei uns tatsächlich komplett leer zum arbeiten, selbst der Papierkorb ist in der STM : )
In den Ordner 'TrayMenu' sollen erstmal alle Ordner Dateien die sehr wichtig sind. Dieser werden immer angezeigt.
Anschließend kann man über 'Einstellungen', dann Tabpage 'Ordner', über die Funktion 'Ordner hinzufügen', den Inhalt von Ordnern wie z.b. den Inhalt vom Ordner 'PRG', hinzufügen. Dort könnte man dann wählen 'Nur Dateien', das bedeutet, dass nur Dateien aus den hinzugefügten Ordner verwendet werden.
Man kann auch den Inhalt vom Desktop auf die gleiche Weise ins root menu bringen. 

In #284 werden wir dann noch eine Spalte 'Nur Suche' implementieren, das bedeutet dann, dass die Inhalte von dem hinzugefügten Ordner nur angezeigt werden, wenn man etwas in der Suche eingibt.

Mit #284 ließe sich dann denken wir konfigurieren, was du möchtest.
Siehst du das auch so? 

Zögere nicht zu fragen wenn etwas unklar ist. Wir freuen uns sehr dich in der Community zu haben. 
Danke für dein Kommentar auf heise : ) !!

(du kannst auf deutsch antworten, ich packe hier noch die Übersetzung von google translate rein)


----

Thanks for your answer!
We are not quite sure if we understand everything and we see you come from Germany, so we formulate it in German.
We see on the screenshot, your root folder is 'TrayMenu'. First of all, everything is currently displayed, files and folders that are in the 'TrayMenu' folder.
The 'TrayMenu' folder can be on the desktop, but I personally have my root folder under D: \ Seafile \ Tools, since Seafile is my cloud, so everything is backed up over the network. The desktop with us is actually completely empty to work, even the paper body is in the STM:)
In the 'TrayMenu' folder, all the files that are very important should first be put into the folder 'TrayMenu'. These are always displayed.
Then you can use 'Settings', then tab page 'Folder', via the 'Add folder' function, the contents of folders such as e.g. add the contents of the folder 'PRG'. There you could then select 'Files only', which means that only files from the added folder will be used.
You can also bring the contents of the desktop into the root menu in the same way.

In # 284 we will implement a column 'Search only', which means that the contents of the added folder will only be displayed if you enter something in the search.

With # 284 we think we can configure what you want.
Do you think so too?

Don't hesitate to ask if something is unclear. We are very happy to have you in the community.
Thanks for your comment on heise :) !!

**Comment by RuSieg** [comment]
@Hofknecht 
Hallo Markus,
Danke für die Möglichkeit in Deutsch zu kommunizieren.
Zunächst, der Ordner ´TrayMenu´ liegt bei mir hinter ´C:\Users\klaus\AppData\Roaming´. In diesen Ordner habe ich die Ordner App, Backup, Gesundheit, Info, Katzen, NAS, Office, Photo-Video, PRG und System, teilweise mit Unterordnern, angelegt. In diesen Ordnern befinden sich die Links zu den entsprechenden Programmen. Im Einzelfall auch zu regulären Verzeichnissen. Diese Struktur soll es mir ermöglichen, a: die entsprechenden Programme nach je nach Aufgabe zu finden und b: überhaupt eine Struktur in meine vielen Programme zu bringen und c: mich überhaupt daran zu erinnern das ich Programme für eine bestimmte Aufgabe installiert habe deren Name mir gar nicht mehr geläufig ist weil ich sie zu selten benutze.
Ich denke da nutze ich SystemTrayMenu auf eine Art wie ihr sie wohl nicht ganz so angedacht habt.
Auch ich habe, dank SystemTrayMenu, einen (fast) leeren Desktop.
Mein Problem: Unterhalb dieser Ordner wird nun Etwas angezeigt das in meinem Fall nur stört und dessen Sinn ich auch nicht verstehe, denn nutzen kann ich es nicht. Es sind offensichtlich Hinweise auf installierte Programme, z.B. OAMEI Backupper. Unter Eigenschaften steht dort richtig "C:\Program Files (x86)\AOMEI\AOMEI Backupper\ABLaucher.exe". Es ist ein Link auf das entsprechende Programm, aber ein Klick oder Doppelklick führt nicht zu dessen Ausführung.
Folgende Wünsche hätte ich nun gerne für meinen, möglicherweise speziellen Gebrauch:
1.	Je nach Einstellung werden diese Links angezeigt oder ausgeblendet.
2.	Wenn sie angezeigt werden sollten die auch ausgeführt werden können.
3.	Wenn schon dieser Link vorhanden ist wäre es schön, wenn ich ihn per Rechtsklick in einen meiner Ordner hinter TrayMenu, als Link, verschieben/kopieren könnte oder ihn wenigstens auf dem Desktop ablegen könnte, von dem ich ihn dann verschieben kann.
Vielen Dank auch für die freundliche Aufnahme und die Bereitschaft auf meine Wünsche einzugehen.
Ich denke im Laufe des täglichen Gebrauchs meinerseits werden mir noch mehr Anregungen einfallen.
Gruß Klaus

**Reply by topeterk** [reply]
Hi,
ich verstehe das so, dass sich diese zusätzlichen Einträge nicht in "C:\Users\klaus\AppData\Roaming\TrayMenu" befinden, korrekt?
Kann es sein, dass die durch Einträge in "Settings -> Folders" eingeblendet werden und das möglicherweise genau die über diese Funktion eingeblendeten Verknüpfungen nicht korrekt funktionieren?
Gruß,
  Peter

**Comment by RuSieg** [comment]
@topeterk 
Hallo,
ja Peter, genau so ist es. Nachdem ich die Einträge entfernt und einem Neustart veranlasst habe, verhält sich SystemTrayMenu so wie ich es gewünscht habe. Danke. Nun kann ich es wie gewünscht verwenden und mich ein wenig tiefer damit befassen. 
Da habe ich offensichtlich das mit den Verzeichniseinträgen falsch verstanden.
Kannst du mir das noch genauer erläutern?

Wenn ich weitere Anregungen habe darf ich mich dann melden?

Gruß Klaus

**Reply by topeterk** [reply]
Hi Klaus,
ja, schön zu hören, dass es jetzt klappt.

Zu dieser Liste gibts hier einige Infos: https://github.com/Hofknecht/SystemTrayMenu/issues/211
Ich meine, es war eine Verbesserung für eine "Schnelle Suche", weil man beim ersten Menü dann direkt in der Suche text eingeben kann und damit direkt auch Einträge "aus anderen Ordnern" (also denen in dieser Liste) finden kann.

Wenn du sonst noch Fragen oder Anregungen hast, gerne einfach melden.
Jede Frage kann zur Verbesserung der Software führen, daher sind alle Fragen jederzeit gern willkommen.

Gruß,
  Peter

**Reply by Tanjalibertatis** [reply]
Hi @RuSieg,

tritt das immer noch auf, wenn du auf eine per "Add Folder" hinzugefügte Verlinkung zu einem Programm klickst/startest, dass es dann nicht geöffnet wird? 
Tritt das bei jeder Verlinkung auf?
Und auch, wenn du die exakt gleiche Verlinkung in den direkten Root-Ordner packst (ohne die Funktion "Add Folder")?

Wenn du willst, kannst du uns davon ein Video machen (z.B. mit "Kostenloser Bildschirmrekorder").
Wenn das Starten eines Prozesses fehlschlägt, sollte anschließend im Logfile eine Zeile dazu stehen, die könntest du uns auch gern schicken, falls dort etwas steht.

Gruß,
Tanja

**Comment by RuSieg** [comment]
@Tanjalibertatis 
Hallo [Tanja,
habe versucht deiner Bitte zu entsprechen. Ergebnis schon eigenartig.
Hier sind 2 Screen-Videos.
Gerne melde ich mich wieder.

Gruß Klaus
[Aufnahme_2022_01_06_16_48_52_520.zip](https://github.com/Hofknecht/SystemTrayMenu/files/7823341/Aufnahme_2022_01_06_16_48_52_520.zip)

**Comment by RuSieg** [comment]
@Tanjalibertatis 
Hallo Tanja,
hier noch das Log.
[log-RUSIEG.zip](https://github.com/Hofknecht/SystemTrayMenu/files/7823709/log-RUSIEG.zip)
Gruß Klaus

**Reply by Hofknecht** [reply]
Wir sehen im logfile: System.UnauthorizedAccessException: Zugriff verweigert (0x80070005 (E_ACCESSDENIED))
Wir können das aber auch mit einem Standardnutzer (nicht Admin) nicht nachstellen.
Es wäre vielleicht ganz interessant, ob es klappt, wenn man die STM.exe mit Adminrechten startet:
![2022-01-06 21_15_50-net6 0-windows10 0 22000 0](https://user-images.githubusercontent.com/52528841/148445851-792264f5-d7d2-4c5a-9ba8-d12609bee6ec.png)
Macht das einen Unterschied?

Ist es bei jedem item so oder passiert das nur bei bestimmten?

(evt können wir im code was ändern: StartInfo LoadUserProfile = true, Verb = "runas", da wir aber nicht nachstellen können, müssten wir eine testversion bauen und dich probieren lassen. : )

**Comment by RuSieg** [comment]
@Hofknecht 
Habe SystemTrayMenu als Administrator gestartet, kein anderes Verhalten, siehe kommentiertes Log.
Dies war über BeispielOrdner
[log-RUSIEG - Kopie.zip](https://github.com/Hofknecht/SystemTrayMenu/files/7827554/log-RUSIEG.-.Kopie.zip)

Nun einen Ordner in dem ein Programm von mir liegt auf D:\ kopiert.
Diesen Ordner aufgenommen, kein Fehler.
Nun einen beliebigen anderen Ordner aufgenommen, kein Fehler, siehe kommentiertes Log.
[log-RUSIEG_3.zip](https://github.com/Hofknecht/SystemTrayMenu/files/7827662/log-RUSIEG_3.zip)
Gruß Klaus

**Comment by Hofknecht** [comment]
@RuSieg
Vielen Dank für die Logs und die Kommentare in den Logs!

Im Log sehen wir nur die Warning, dass es die System.UnauthorizedAccessException beim Lesen des '...\Start Menu\Programme' gibt.
Das überspringt er dann. Dieser Ordner wurde über Einstellungen - Folders - hinzugefügt.
Wir sind jetzt gerade nicht sicher, ob er die nicht lesbare (keine Rechte) Konfiguration in dem Moment korrekt überspringt und alle weiteren dann korrekt macht. Wenn du die fehlerhafte (keine Rechte) Konfiguration entfernst, dann funktioniert alles richtig, oder?

Die Frage ist natürlich schon warum auf diesen Ordner keine Rechte sind und ob wir daran etwas ändern können. Wir haben das Thema mit "c# GetDirectories UnauthorizedAccessException" gegoogelt. Zu dem Thema ist eventuell das ganz interessant: http://www.blackwasp.co.uk/FolderRecursion.aspx
Da sollte man also vielleicht was machen. Wir sind uns jetzt nicht sicher, ob wir die Situation komplett verstehen.
Da wir gerade die Situation noch nicht nachstellen können, wäre es natürlich perfekt, wenn du probieren magst, ob die Situation bei dir nochmal nachstellbar und reproduzierbar ist und das müsste man anhand dessen dann nochmal korrekt durchdenken.

Wir sehen du hast die Version 1.2.0.9 verwendet. Das bedeutet wahrscheinlich, dass du Visual Studio 2022 drauf hast und dir eine Version bauen kannst. Vielleicht hast du ja sogar Lust, dir die Stelle im Code anzusehen bzw. dort zu debuggen. Wir würden dich natürlich bestmöglich unterstützen.

Aus dem Logfile sehen wir, es tritt hier auf:
> 2022.01.07 08:20:23.267 .NET ThreadPool Worker  WARN GetDirectoriesAndFilesRecursive path:'C:\ProgramData\Microsoft\Windows\Start Menu\Programme' System.UnauthorizedAccessException: Access to the path 'C:\ProgramData\Microsoft\Windows\Start Menu\Programme' is denied.    at System.IO.Enumeration.FileSystemEnumerator`1.CreateDirectoryHandle(String path, Boolean ignoreNotFound)    at System.IO.Enumeration.FileSystemEnumerator`1.Init()    at System.IO.Enumeration.FileSystemEnumerator`1..ctor(String directory, Boolean isNormalized, EnumerationOptions options)    at System.IO.Enumeration.FileSystemEnumerable`1..ctor(String directory, FindTransform transform, EnumerationOptions options, Boolean isNormalized)    at System.IO.Enumeration.FileSystemEnumerableFactory.UserDirectories(String directory, String expression, EnumerationOptions options)    at System.IO.Directory.InternalEnumeratePaths(String path, String searchPattern, SearchTarget searchTarget, EnumerationOptions options)    at System.IO.Directory.GetDirectories(String path, String searchPattern, EnumerationOptions enumerationOptions)    at System.IO.Directory.GetDirectories(String path)    at SystemTrayMenu.Business.Menus.GetDirectoriesAndFilesRecursive(String[]& directoriesToConcat, String[]& filesToAddToConcat, String pathAddToMainMenu) in D:\Data\VS\Api's\Fremde\TrayMenu\SystemTrayMenu-master\Business\Menus.cs:line 767

also bei der Methode GetDirectoriesAndFilesRecursive
(es könnte jetzt aber auch sein, das die warning nichts mit dem problem zu tun hat. Tritt es speziell auf, wenn du einen bestimmten Ordner dort hinzufügt bzw hängt das mit dem recursiv flag auch zusammen?)

Aber das wichtigste ist, dass es ohne fehlerhafte (keine Rechte) Konfiguration funktioniert. Ist das so korrekt?
Funktioniert sonst alles mit zusätzliche Ordner hinzufügen und starten (Klick / Doppelklick) von Dateien und Ordnern usw.?
Also gibt es einen großen Blocker? 

Falls das hohe Prio bei dir hat, dann leg gerne dafür ein Issue an und wir fassen das da zusammen oder wir können das auch hier noch weiter besprechen. Wir freuen uns über jede Erkenntnis.
Falls aus der Erfahrung hier nochmal etwas für andere Benutzer interessant sein könnte, dann wäre es wahrscheinlich auch richtig, ein Issue anzulegen und dort die Zusammenfassung (google translate auf english) für alle zu teilen. Das könnten wir dann auch als FAQ markieren und closen, oder falls es noch offen ist, mit help wanted makieren.


Hoffentlich war das nicht zu viel und wir überfordern dich nicht : )
Wir hoffen du kannst den Fall nochmal reproduzieren und uns da noch weiter bringen oder das es kein großes Problem ist.
Es ist auch hilfreich das gesamte problem nochmal in kürze zusammen zu fassen bzw es auf einen kleinstmöglichsten reproduzierbaren Testcase zu bringen. Alles hilft weiter!
Vielen Dank für deine Unterstützung !!!

**Comment by Hofknecht** [comment]
oh ok we just noticed that it also appears in our main menu.
it probably has nothing to do with the exception.
We'll look at it again during the next few days and let you know, if this happens often we'll bring out a hotfix next week

**Comment by Hofknecht** [comment]
Hey @RuSieg 
in my case it was related to #275, now fixed with version 1.2.1.1 (see commit: 0511468) (unfortunately not in release 1.2.1)
can you build version 1.2.1.1 and tell us if this solved the issue for you? 
Thank you in advanced!


(added also a issue/ bug, that we see it in the next release notes:  https://github.com/Hofknecht/SystemTrayMenu/issues/289)

**Comment by RuSieg** [comment]
Ich melde mich am Montag.
--
Gesendet über myMail für Android Samstag, 08 Januar 2022, 11:27PM +01:00 von Markus Hofknecht  ***@***.*** :

**Comment by RuSieg** [comment]
@Hofknecht 
Hallo Markus,
habe die Version 1.2.2.0 mit Visual Studio 2022 auf Windows 11 kompiliert und gestartet, ohne Fehlermeldung.
Danach in meinem Menü ein Programm gestartet, klappte einwandfrei, aber danach Fehlermeldungen, siehe Protokoll.
[log-RUSIEG_2022.01.10.001.zip](https://github.com/Hofknecht/SystemTrayMenu/files/7837841/log-RUSIEG_2022.01.10.001.zip)
Es sind aber nicht alle Einträge in meinen Ordnern
Zuerst dachte ich, das die Fehlermeldungen nur vorkommen wenn ich ein Programm starte dass Adminrechte erwartet.
Dem war aber nicht so, ich kann keine Systematik erkennen.

Selbstverständlich habe ich Lust mir die Stelle im Code anzusehen und zu debuggen. Hier brauche ich aber eure Unterstützung.
Gruß Klaus

**Comment by Hofknecht** [comment]
@RuSieg 
Hey Klaus,

Ich sehe grad 3 Themen:

1.
ok sehr gut zu hören, das der Fehler mit dem "klick funktioniert nicht" erstmal gelöst ist mit https://github.com/Hofknecht/SystemTrayMenu/issues/289
Die Version 1.2.2.0 gibt es jetzt auch als release: https://github.com/Hofknecht/SystemTrayMenu/releases
(release 1.2.0.0 passte noch, 1.2.1.0 hat den Fehler)

2.
Die Warnings in deinem logfile passieren hier:
![2022-01-10 17_14_19-Window](https://user-images.githubusercontent.com/52528841/148801679-d2ac56f0-0138-4f27-b871-4e4be199777d.png)
Die funktion ist aktuell nur noch wichtig um zu ermitteln, ob es sich beim ziel vom shortcut (*.LNK) um einen Ordner handelt. 
Bei dir sind es jetzt tatsächlich nur Dateien, daher kein Problem. Bei einem Ordner könnte er diesen nicht auflösen.
Ich überlege ob wir die Warning irgendwie nur loggen wenn es ein Ordner ist, genau das wissen wir dann aber noch nicht.
man findet dazu in google sowas: https://stackoverflow.com/questions/1395205/better-way-to-check-if-a-path-is-a-file-or-a-directory 
wenn ich mich recht errinere hatten wir hier jedoch andere probleme, z.b. ging es im netzwerk nicht.
=> Daher denke ich gerade die Warnings im logfile können erstmal so bleiben

3.
"Es sind aber nicht alle Einträge in meinen Ordnern"
Das sollten wir nochmal genau anschauen. 
Ist das fehlende Item ein Ordner oder Datei, direkt im Haupt-Ordner oder über einen Ordner hinzugefügt?
Ist es sichtbar z.b. wenn du es direkt in den Haupt-Ordner packst?
Ist es im Untermenu sichtbar z.b. wenn du es direkt in einen Ordner im Haupt-Ordner packst?
Ist es im Hauptmenu sichtbar, wenn du diesen unterordner einbindest über einstellungen Tabpage 'Ordner'?
(Achso oder du meinst gar nicht das ein eintrag fehlt, sondern einfach die warning nur ein paar items betrifft? dann passt natürlich alles und wir können diese discussion erstmal closen, aber öffne gerne weitere, wir freuen uns über weitere Themen : )

**Comment by RuSieg** [comment]
@Hofknecht 
Hallo Markus,
es tut mir leid, ich kann keine Systematik erkennen.
z.B. Ich lösche das Log, rufe das Menü ab und fahre mit dem Cursor über die Einträge ohne das ich die eine Maustaste nutze.
Manchmal wird plötzlich wird die Logdatei angelegt und enthält Einträge. Wenn ich dem systematisch nachgehe indem ich bewusst die einzelnen Ordner Einträge öffne, bis in die letzte Ebene, passiert gar nichts. Manchmal aber eben doch. Ich bin ratlos.

**Comment by Hofknecht** [comment]
@RuSieg 
Hey Klaus,
solange sich alles richtig verhält, ist es kein Problem. Die Warnings im logfile sind soweit ok.

**Comment by RuSieg** [comment]
@Hofknecht 
Hallo Markus,
ja ich denke dieses Thema kann endgültig abgeschlossen werden.
Ich kann SystemTrayMenu wie gewünscht benutzen, danke noch einmal für die Unterstützung.
Gruß Klaus
